src/dataflow/core/type_inference.py

```diff
@@ -0,0 +1,330 @@
+        ascribed_return_type = (
+            _type_name_to_type(expression.type, {})
+            if expression.type
+            else AnonTypeVariable()
+        )
+
+        if expression.type_args:
+            for (ascribed_type_arg, type_var) in zip(
+                expression.type_args, declared_type_args_list
+            ):
+                substitutions[type_var] = _type_name_to_type(ascribed_type_arg, {})
+        ascribed_arg_types = (
+            [cast(Type, AnonTypeVariable()) for arg in defn.args] if defn else []
+        )
+        ascribed_type = TypeApplication(
+            "Lambda", ascribed_arg_types + [ascribed_return_type]
+        )
+        comp_type = _unify(ascribed_type, defn_type, substitutions)
+        assert isinstance(
+            comp_type, TypeApplication
+        ), "unification of Lambdas should always produce a TypeApplication"
+        return Computation(
+            op, expression.id, rec_args, declared_type_args_list, comp_type
+        )
+
+    (roots, _) = roots_and_reentrancies(program)
+    assert (
+        len(roots) == 1
+    ), f"Expected Program to have a single root, got {roots} in {program}"
+    root_id = list(roots)[0]
+    return rec(id_to_expr[root_id])
+
+
+def _definition_to_type(
+    definition: Definition, declared_type_args: Dict[str, NamedTypeVariable]
+) -> Type:
+    return_type = _type_name_to_type(definition.type, declared_type_args)
+    arg_types = [_type_name_to_type(arg, declared_type_args) for arg in definition.args]
+
+    return TypeApplication("Lambda", arg_types + [return_type])
+
+
+def _type_name_to_type(
+    name: TypeName, declared_type_args: Dict[str, NamedTypeVariable]
+) -> Type:
+    if name is None:
+        return AnonTypeVariable()
+    if len(name.type_args) == 0:
+        if name.base in declared_type_args:
+            return declared_type_args[name.base]
+        else:
+            return TypeApplication(name.base, [])
+    else:
+        args = [_type_name_to_type(arg, declared_type_args) for arg in name.type_args]
+        return TypeApplication(name.base, args)
+
+
+def _type_to_type_name(t: Type) -> TypeName:
+    assert isinstance(
+        t, TypeApplication
+    ), f"_type_to_type_name expects a Type that is fully instantiated with no variables, but got {t}"
+    return TypeName(t.constructor, [_type_to_type_name(a) for a in t.args])
+
+
+def _apply_substitutions(t: Type, substitutions: Dict[TypeVariable, Type]):
+    if isinstance(t, TypeVariable):
+        if t in substitutions:
+            return _apply_substitutions(substitutions[t], substitutions)
+        else:
+            return t
+    elif isinstance(t, TypeApplication):
+        return TypeApplication(
+            t.constructor, [_apply_substitutions(arg, substitutions) for arg in t.args]
+        )
+
+    else:
+        raise TypeInferenceError(f"Unknown type {t}")
+
+
+def _unify(t1: Type, t2: Type, substitutions: Dict[TypeVariable, Type]) -> Type:
+    """See e.g. section 2.2 of
```

Comment:
Document the `Returns ` and the `Raises`.
(see [this example](https://google.github.io/styleguide/pyguide.html#383-functions-and-methods).)

Reply:
Since this function has the least self-documenting name (unless you're familiar with type inference algorithms), I'd add a bit more to the doc string - at the very least describe the contract. For example, it'd be good to know why we pass in `substitutions` rather than simply applying it to `t1` and `t2` before passing those in.